src/enumerable.cr

```diff
@@ -1747,11 +1747,29 @@ module Enumerable(T)
   # ```
   # ["a", "A", "b", "B"].tally_by(&.downcase) # => {"a" => 2, "b" => 2}
   # ```
-  def tally_by(& : T -> U) : Hash(U, Int32) forall U
-    each_with_object(Hash(U, Int32).new) do |item, hash|
+  def tally_by(&block : T -> U) : Hash(U, Int32) forall U
+    tally_by(Hash(U, Int32).new, &block)
+  end
+
+  # Tallies the collection. Accepts a `hash` to count occurrences.
+  # The value corresponding to each element must be an integer.
+  # Returns a `hash` where the keys are the
+  # elements and the values are numbers of elements in the collection
+  # that correspond to the key after transformation by the given block.
+  #
+  # ```
+  # hash = {} of Char => Int32
+  # words = ["Crystal", "Ruby"]
+  # words.each { |word| word.chars.tally_by(hash, &.downcase) }
+  # hash
+  # => {'c' => 1, 'r' => 2, 'y' => 2, 's' => 1, 't' => 1, 'a' => 1, 'l' => 1, 'u' => 1, 'b' => 1}
+  # ```
+  def tally_by(hash)
+    each_with_object(hash) do |item, hash|
       value = yield item
-      count = hash[value]?
-      hash[value] = count ? count + 1 : 1
+
+      count = hash.has_key?(value) ? hash[value] : typeof(hash[value]?.not_nil!).zero
```

Comment:
💭  This is doing a double hash lookup, so I'm thinking it could be written like this:

```crystal
existing_count = hash[value]?
count = existing_count || typeof(hash[value]).zero
```

Or shorter like:

```crystal
count = hash[value]? || typeof(hash[value]).zero
```

Or even:

```crystal
count = hash.fetch(value) { typeof(hash[value]).zero }
```

Maybe this last one my preferred option.

Reply:
My vote is the last one

```
count = hash.fetch(value) { typeof(hash[value]).zero }
```

Reply:
This is awesome :pray: